### ISSUE TITLE
Venv resolution conflict with `./.venv` and `.python-version` file

### ISSUE THREAD

**Issue by systemsoverload** [sublimelsp/LSP-pyright#331]
## Expected:
Projects with a `.venv` should always have python loaded from this directory.

## Actual:
Projects that have a `.venv` directory in the base of the project as well as a `.python-version` file in the base of project default to the command `pyenv which python {version from .python-version}` to resolve the venv and ignore the local venv. 

This layout is the default behavior of [Rye](https://rye.astral.sh/) managed projects.

**Comment by jfcherng** [comment]
Fwiw, we basically ported https://github.com/fannheyward/coc-pyright/blob/6011417adce4098438a66883a2d03bea4a4d68d7/src/configSettings.ts#L47-L114.

**Comment by catwell** [comment]
I usse Rye too and have the same issue. The main issue here is with [this commit](https://github.com/sublimelsp/LSP-basedpyright/commit/bbe02911466c660590211fbd5741ef8728cde000) you do not check the return code of the subprocess anymore!

So before (with `check_output`) it raised an exception, which was caught, and it continued to the venv detection code.

Now it fails silently, so it triggers the return:

![Screenshot from 2024-06-03 16-54-02](https://github.com/sublimelsp/LSP-pyright/assets/221332/33219422-06c1-48f5-9782-354ce9075bf7)

**Comment by jfcherng** [comment]
> I use Rye too and have the same issue. The main issue here is with [this commit](https://github.com/sublimelsp/LSP-basedpyright/commit/bbe02911466c660590211fbd5741ef8728cde000) you do not check the return code of the subprocess anymore!

That commit is even after this issue was filed.

**Comment by jfcherng** [comment]
> Fwiw, we basically ported [fannheyward/coc-pyright@`6011417`/src/configSettings.ts#L47-L114](https://github.com/fannheyward/coc-pyright/blob/6011417adce4098438a66883a2d03bea4a4d68d7/src/configSettings.ts#L47-L114). But I do think your proposal makes more sense.

As I stated.

**Comment by catwell** [comment]
True! So my issue is slightly different since in my case it worked before this commit, i.e. before the last release (I use Rye but I do not have pyenv installed). But fixing this one would solve my issue too.

**Comment by jfcherng** [comment]
I guess because https://github.com/fannheyward/coc-pyright/blob/6011417adce4098438a66883a2d03bea4a4d68d7/src/configSettings.ts#L54-L57 so coc.vim works.

**Comment by jfcherng** [comment]
I guess should add `.venv` folder as a special case (for better UX) as the highest priority. 

I think it makes sense to search "arbitrary" subfolders to find a virtual env as the last help as in https://github.com/sublimelsp/LSP-pyright/blob/7d9269b0400eb2129e351f3a63f80f6704f4409c/plugin.py#L245-L251.

---

The strategy coc-pyright used now is basically rely on env variables. Unfortunately that won't work well in ST since env variables are the same for all windows and last until ST gets closed.

**Comment by jfcherng** [comment]
I am thinking how about let the user able to decide the search order 🤔 

Something like

```js
    "venv_search_order": [
        "dot_venv",
        "dot_pdm_python",
        "dot_python_version",
        "pipfile",
        "poetry_lock",
        "any_subfolder",
    ],
```

**Comment by catwell** [comment]
That works too! Actually a config like this is the first thing I looked for.

**Comment by jfcherng** [comment]
> That works too! Actually a config like this is the first thing I looked for.

In the next release (probably the day after tomorrow), you can decide how venv is resolved.

By default, `.venv` at the project root is preferred. https://github.com/sublimelsp/LSP-pyright/blob/bab20e63ab45e6eaf317fe07cef088f1e478e50e/LSP-pyright.sublime-settings#L20-L32

**Comment by jfcherng** [comment]
Update: I change the settings name from `pyright.venv.strategies` to `venvStrategies`.

**Comment by jfcherng** [comment]
[1.4.0](https://github.com/sublimelsp/LSP-pyright/releases/tag/1.4.0) has been released and should be available within hours. You may need some ST restarts to make it work since new lib dependencies are added.
